runtime/src/main/java/org/corfudb/protocols/wireprotocol/StreamAddressSpace.java

```diff
@@ -0,0 +1,154 @@
+package org.corfudb.protocols.wireprotocol;
+
+import org.corfudb.runtime.view.Address;
+import org.roaringbitmap.longlong.LongIterator;
+import org.roaringbitmap.longlong.Roaring64NavigableMap;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.NavigableSet;
+import java.util.TreeSet;
+
+/**
+ * This class represents the space of all addresses belonging to a stream.
+ *
+ * A stream's address space is defined by:
+ *       1. The collection of all addresses that belong to this stream.
+ *       2. The trim mark (last trimmed address, i.e., an address that is no longer present and that was subsumed by
+ *       a checkpoint).
+  *
+ * Created by annym on 03/06/2019
+ */
+public class StreamAddressSpace {
+
+    private static final int NO_ADDRESSES = 0;
+
+    // Holds the last trimmed address for this stream.
+    // Note: keeping the last trimmed address is required in order to properly set the stream tail on sequencer resets
+    // when a stream has been checkpointed and trimmed and there are no further updates to this stream.
+    private long trimMark;
+
+    // Holds the complete map of addresses for this stream.
+    private final Roaring64NavigableMap addressMap;
+
+    public StreamAddressSpace(long trimMark, Roaring64NavigableMap addressMap) {
+        this.trimMark = trimMark;
+        this.addressMap = addressMap;
+    }
+
+    public Roaring64NavigableMap getAddressMap() {
+        return addressMap;
+    }
+
+    /**
+     * Copy this stream's addresses to a set, under a given boundary (inclusive).
+     *
+     * @param maxGlobal maximum address (inclusive upper bound)
+     */
+    public NavigableSet<Long> copyAddressesToSet(final Long maxGlobal) {
+        NavigableSet<Long> queue = new TreeSet<>();
+        this.addressMap.forEach(address -> {
+            if (address <= maxGlobal){
+                queue.add(address);
+            }
+        });
+
+        return queue;
+    }
+
+    /**
+     * Checks if this stream has addresses or not.
+     *
+     * @return true, if no addresses belong to this stream; false, otherwise.
+     */
+    public boolean isEmpty() {
+       return this.addressMap.getLongCardinality() == NO_ADDRESSES;
+    }
+
+    /**
+     * Get tail for this stream.
+     *
+     * @return last address belonging to this stream
+     */
+    public Long getTail() {
+        // If no address is present for this stream, the tail is given by the trim mark (last trimmed address)
+        if (addressMap.getLongCardinality() == NO_ADDRESSES) {
+            return trimMark;
+        } else {
+            // The stream tail is the max address present in the stream's address map
+            return this.addressMap.getReverseLongIterator().next();
+        }
+    }
+
+    /**
+     * Add an address to this address space.
+     *
+     * @param address address to add.
+     */
+    public void addAddress(long address) {
+        this.addressMap.addLong(address);
```

Comment:
please get rid of all extra `this` everywhere in yours pr. It's boilerplate code.